### ISSUE TITLE
Credentials not set after master takeover

### ISSUE THREAD

**Issue by d-rupp** [openark/orchestrator#1377]
This issue was introduced between 3.2.3 (working) and 3.2.5 (broken).

Setup:
The replication credentials are stored in a  metadata table on the database server, Orchestrator knows about these with ReplicationCredentialsQuery. 

In 3.2.3 - issue graceful-master-takeover and orchestrator automatically connects the old master to the new one.

In 3.2.5 - after graceful-master-takeover the old master is set as slave but username is missing (password possibly too). 

```
Last IO error | "Fatal error: Invalid (empty) username when attempting to connect to the master server. Connection attempt terminated."
```

**Comment by shlomi-noach** [comment]
You're right. The change was introduced in https://github.com/openark/orchestrator/pull/1250. It was noted a few days ago in https://github.com/openark/orchestrator/pull/1375

I'll open a PR to fix the code to make it both backwards and forwards compatible.

**Comment by shlomi-noach** [comment]
Fixed in https://github.com/openark/orchestrator/pull/1378. Binary artifact is in https://github.com/openark/orchestrator/suites/3110208007/artifacts/71167636

Are you able to test this?

**Comment by d-rupp** [comment]
Yes! I deployed this in our development environment and it works like a charm! Thanks!

**Comment by shlomi-noach** [comment]
Cool, thanks! Reopening since the or is not merged yet.

**Comment by d-rupp** [comment]
Whops, sorry :)

**Comment by shlomi-noach** [comment]
#1378 now merged

**Comment by leiless** [comment]
Hi @shlomi-noach, I've also met this problem today after I manually promote a slave to master(via graceful-master-takeover):

Using `10.5.10-MariaDB-1:10.5.10+maria~focal-log` and `orchestrator 3.2.5`

```
Slave_IO_State:
Master_Host: mariadb-13307
Master_User: 
Master_Port: 3306
Last_IO_Error: Fatal error: Invalid (empty) username when attempting to connect to the master server. Connection attempt terminated.
```

**Comment by shlomi-noach** [comment]
@leiless cool. Want to try out the fix?

**Comment by leiless** [comment]
> @leiless cool. Want to try out the fix?

Seems not working, I've checked twice.
I downloaded the `orchestrator.zip` from https://github.com/openark/orchestrator/suites/3110208007/artifacts/71167636 and replaced `/usr/local/orchestrator/orchestrator`

FYI:
```bash
$ ls -l orchestrator.zip
-rw-rw-r-- 1 lei lei 11006501 Jun 30 15:52 orchestrator.zip

$ shasum -a 256 orchestrator.zip 
93cadd0c489bd643ec1ed298b4263e902550dc59917a8d48e3e9064440a16927  orchestrator.zip

$ ls -l /usr/local/orchestrator/orchestrator
-rwxrwxr-x 1 lei lei 21545904 Jun 29 06:37 /usr/local/orchestrator/orchestrator

$ shasum -a 256 /usr/local/orchestrator/orchestrator                                
6f17153a7642b96e23e7990e3f705b9e354bf96bd3c8476420b11e71fb63b1c1  /usr/local/orchestrator/orchestrator

$ /usr/local/orchestrator/orchestrator -version
de5176436875c415890356855075ea5d0feb5fcd
```

### Steps to reproduce

![image](https://user-images.githubusercontent.com/38041294/123928495-8ac44980-d9c0-11eb-8a9d-8030f7a04106.png)
![image](https://user-images.githubusercontent.com/38041294/123928545-9748a200-d9c0-11eb-9682-6c54bfb57520.png)
### Goto http://127.0.0.1:3000/web/cluster/alias/mariadb-13307
![image](https://user-images.githubusercontent.com/38041294/123928581-9e6fb000-d9c0-11eb-9e80-40209ac0eed3.png)

### Manually _Start replication_ on the `mariadb-13306:3306` node
![image](https://user-images.githubusercontent.com/38041294/123928745-c3642300-d9c0-11eb-94d5-c223da69d62e.png)

### `show slave status` on `mariadb-13306`
```
$ mysql -h172.31.254.1 -uroot -p123456
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 28
Server version: 10.5.10-MariaDB-1:10.5.10+maria~focal-log mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> show slave status \G
*************************** 1. row ***************************
                Slave_IO_State: 
                   Master_Host: mariadb-13307
                   Master_User: 
                   Master_Port: 3306
                 Connect_Retry: 60
               Master_Log_File: 
           Read_Master_Log_Pos: 4
                Relay_Log_File: mysqld-relay-bin.000001
                 Relay_Log_Pos: 4
         Relay_Master_Log_File: 
              Slave_IO_Running: No
             Slave_SQL_Running: Yes
               Replicate_Do_DB: 
           Replicate_Ignore_DB: 
            Replicate_Do_Table: 
        Replicate_Ignore_Table: 
       Replicate_Wild_Do_Table: 
   Replicate_Wild_Ignore_Table: 
                    Last_Errno: 0
                    Last_Error: 
                  Skip_Counter: 0
           Exec_Master_Log_Pos: 4
               Relay_Log_Space: 256
               Until_Condition: None
                Until_Log_File: 
                 Until_Log_Pos: 0
            Master_SSL_Allowed: No
            Master_SSL_CA_File: 
            Master_SSL_CA_Path: 
               Master_SSL_Cert: 
             Master_SSL_Cipher: 
                Master_SSL_Key: 
         Seconds_Behind_Master: NULL
 Master_SSL_Verify_Server_Cert: No
                 Last_IO_Errno: 1593
                 Last_IO_Error: Fatal error: Invalid (empty) username when attempting to connect to the master server. Connection attempt terminated.
                Last_SQL_Errno: 0
                Last_SQL_Error: 
   Replicate_Ignore_Server_Ids: 
              Master_Server_Id: 0
                Master_SSL_Crl: 
            Master_SSL_Crlpath: 
                    Using_Gtid: Current_Pos
                   Gtid_IO_Pos: 1-1-7198
       Replicate_Do_Domain_Ids: 
   Replicate_Ignore_Domain_Ids: 
                 Parallel_Mode: optimistic
                     SQL_Delay: 0
           SQL_Remaining_Delay: NULL
       Slave_SQL_Running_State: Slave has read all relay log; waiting for more updates
              Slave_DDL_Groups: 0
Slave_Non_Transactional_Groups: 0
    Slave_Transactional_Groups: 0
1 row in set (0.001 sec)
```

**Comment by leiless** [comment]
`sudo ./orchestrator --debug --stack --config /etc/orchestrator.conf.json http`

<details>
<summary>Operation logs</summary>

```
2021-06-30 16:23:23 INFO auditType:forget-clustr-aliases instance::0 cluster: message:Forgotten aliases: 0
2021-06-30 16:23:23 INFO auditType:forget-unseen instance::0 cluster: message:Forgotten instances: 0
2021-06-30 16:23:23 INFO auditType:resolve-unknown-masters instance::0 cluster: message:Num resolved hostnames: 0
2021-06-30 16:23:23 INFO auditType:review-unseen-instances instance::0 cluster: message:Operations: 0
2021-06-30 16:23:23 DEBUG kv.SubmitMastersToKvStores, clusterName: , force: false: numPairs: 5
2021-06-30 16:23:23 INFO auditType:forget-unseen-differently-resolved instance::0 cluster: message:Forgotten instances: 0
2021-06-30 16:23:23 INFO auditType:inject-unseen-masters instance::0 cluster: message:Operations: 0
2021-06-30 16:23:23 DEBUG kv.SubmitMastersToKvStores: submitKvPairs: 0
[martini] Started GET /api/graceful-master-takeover/mariadb-13306/3306/mariadb-13307/3306 for 127.0.0.1:59466
2021-06-30 16:23:23 INFO GracefulMasterTakeover: designated master instructed to be mariadb-13307:3306
2021-06-30 16:23:23 INFO GracefulMasterTakeover: Will let mariadb-13307:3306 take over its siblings
2021-06-30 16:23:23 INFO moveReplicasViaGTID: Will move 1 replicas below mariadb-13307:3306 via GTID, max concurrency: 5
2021-06-30 16:23:23 INFO Will move mariadb-13308:3306 below mariadb-13307:3306 via GTID
2021-06-30 16:23:23 INFO auditType:begin-maintenance instance:mariadb-13308:3306 cluster:mariadb-13306:3306 message:maintenanceToken: 1, owner: lei-thinkcentre-e12, reason: move below mariadb-13307:3306
2021-06-30 16:23:23 INFO Stopped replication on mariadb-13308:3306, Self:mysql-bin.000003:342, Exec:mysql-bin.000003:342
2021-06-30 16:23:23 DEBUG ChangeMasterTo: will attempt changing master on mariadb-13308:3306 to mariadb-13307:3306, mysql-bin.000003:342
2021-06-30 16:23:23 INFO ChangeMasterTo: Changed master on mariadb-13308:3306 to: mariadb-13307:3306, mysql-bin.000003:342. GTID: true
2021-06-30 16:23:23 INFO Started replication on mariadb-13308:3306
2021-06-30 16:23:23 INFO auditType:move-below-gtid instance:mariadb-13308:3306 cluster:mariadb-13306:3306 message:moved mariadb-13308:3306 below mariadb-13307:3306
2021-06-30 16:23:23 INFO auditType:end-maintenance instance:mariadb-13308:3306 cluster:mariadb-13306:3306 message:maintenanceToken: 1
2021-06-30 16:23:23 INFO auditType:move-replicas-gtid instance:mariadb-13307:3306 cluster:mariadb-13306:3306 message:moved 1/1 replicas below mariadb-13307:3306 via GTID
2021-06-30 16:23:23 INFO auditType:relocate-replicas instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:relocated 1 replicas of mariadb-13306:3306 below mariadb-13307:3306
2021-06-30 16:23:23 INFO GracefulMasterTakeover: Will demote mariadb-13306:3306 and promote mariadb-13307:3306 instead
2021-06-30 16:23:23 ERROR Error 1142: SELECT command denied to user 'orchestrator'@'172.31.254.254' for table 'slave_master_info'
goroutine 6968 [running]:
runtime/debug.Stack(0x81, 0x1800, 0xc000295180)
	/opt/hostedtoolcache/go/1.14.15/x64/src/runtime/debug/stack.go:24 +0x9d
runtime/debug.PrintStack()
	/opt/hostedtoolcache/go/1.14.15/x64/src/runtime/debug/stack.go:16 +0x22
github.com/openark/orchestrator/vendor/github.com/openark/golib/log.logErrorEntry(0x2, 0xd98c20, 0xc000bab040, 0xc000257718, 0x2)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/openark/golib/log/log.go:188 +0xe7
github.com/openark/orchestrator/vendor/github.com/openark/golib/log.Errore(...)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/openark/golib/log/log.go:234
github.com/openark/orchestrator/go/inst.ReadReplicationCredentials(0xc000c0ec00, 0xc8b2bd, 0x3f, 0xc000257be8)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/go/inst/instance_topology_dao.go:1100 +0x10a
github.com/openark/orchestrator/go/logic.GracefulMasterTakeover(0xc000290ac0, 0x12, 0xc000e4e6a0, 0xc000518000, 0x4, 0x1, 0xc000f23580, 0xd)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/go/logic/topology_recovery.go:2093 +0x81f
github.com/openark/orchestrator/go/http.(*HttpAPI).gracefulMasterTakeover(0x126e240, 0xc00043bc20, 0x7fa1ac0ca180, 0xc000215800, 0xc000139700, 0x0, 0x0, 0x41b800)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/go/http/api.go:3178 +0x2a8
github.com/openark/orchestrator/go/http.(*HttpAPI).GracefulMasterTakeover(...)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/go/http/api.go:3194
reflect.Value.call(0xb92860, 0xc000613ac0, 0x13, 0xc5aca4, 0x4, 0xc000f01620, 0x4, 0x4, 0xc000258638, 0x4540fc, ...)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xb92860, 0xc000613ac0, 0x13, 0xc000f01620, 0x4, 0x4, 0x10, 0xbe32e0, 0x1)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:321 +0xb4
github.com/openark/orchestrator/vendor/github.com/codegangsta/inject.(*injector).Invoke(0xc000e4e000, 0xb92860, 0xc000613ac0, 0xb970c0, 0xc00009ce10, 0xc000258768, 0xc00010a3c0, 0x16)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/codegangsta/inject/inject.go:102 +0x2e3
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*routeContext).run(0xc00043bc50)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/router.go:408 +0x73
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*route).Handle(0xc0002b4190, 0x7fa1ac0ca0d0, 0xc0000920a0, 0x7fa1ac0ca150, 0xc000e4e120)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/router.go:285 +0x108
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*router).Handle(0xc000162120, 0x7fa1ac0ca150, 0xc000e4e120, 0xc000139700, 0x7fa1ac0ca0d0, 0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/router.go:132 +0x2e2
reflect.Value.call(0xb84f20, 0xc0004bcde0, 0x13, 0xc5aca4, 0x4, 0xc000092780, 0x3, 0x3, 0xc000258c58, 0x4540fc, ...)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xb84f20, 0xc0004bcde0, 0x13, 0xc000092780, 0x3, 0x3, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:321 +0xb4
github.com/openark/orchestrator/vendor/github.com/codegangsta/inject.(*injector).Invoke(0xc000e4e000, 0xb84f20, 0xc0004bcde0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/codegangsta/inject/inject.go:102 +0x2e3
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).run(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:173 +0x71
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).Next(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:164 +0x2f
github.com/openark/orchestrator/vendor/github.com/martini-contrib/gzip.glob..func1(0x7fa1ac0ca0a0, 0xc000a7e3c0, 0xc000139700, 0x7fa1ac0ca0d0, 0xc0000920a0, 0xffffffffffffffff)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/martini-contrib/gzip/gzip.go:40 +0x385
github.com/openark/orchestrator/vendor/github.com/martini-contrib/gzip.All.func1(0x7fa1ac0ca0a0, 0xc000a7e3c0, 0xc000139700, 0x7fa1ac0ca0d0, 0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/martini-contrib/gzip/gzip.go:56 +0x63
reflect.Value.call(0xb84f20, 0xc000526600, 0x13, 0xc5aca4, 0x4, 0xc000092500, 0x3, 0x3, 0xc000a23168, 0x4540fc, ...)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xb84f20, 0xc000526600, 0x13, 0xc000092500, 0x3, 0x3, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:321 +0xb4
github.com/openark/orchestrator/vendor/github.com/codegangsta/inject.(*injector).Invoke(0xc000e4e000, 0xb84f20, 0xc000526600, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/codegangsta/inject/inject.go:102 +0x2e3
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).run(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:173 +0x71
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).Next(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:164 +0x2f
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.Recovery.func1(0x7fa1ac0ca0d0, 0xc0000920a0, 0xc000093400)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/recovery.go:142 +0x6c
reflect.Value.call(0xb75ba0, 0xcbd0a0, 0x13, 0xc5aca4, 0x4, 0xc00009cf30, 0x2, 0x2, 0xc000a235a8, 0x4540fc, ...)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xb75ba0, 0xcbd0a0, 0x13, 0xc00009cf30, 0x2, 0x2, 0x6e, 0xc0, 0x6e)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:321 +0xb4
github.com/openark/orchestrator/vendor/github.com/codegangsta/inject.(*injector).Invoke(0xc000e4e000, 0xb75ba0, 0xcbd0a0, 0x5096be, 0xc000093400, 0xc000a236d0, 0x412005, 0xbe33a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/codegangsta/inject/inject.go:102 +0x2e3
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).run(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:173 +0x71
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).Next(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:164 +0x2f
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.Logger.func1(0x7fa1ac0ca0a0, 0xc000a7e3c0, 0xc000139700, 0x7fa1ac0ca0d0, 0xc0000920a0, 0xc000093400)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/logger.go:25 +0x210
reflect.Value.call(0xb933a0, 0xcbd088, 0x13, 0xc5aca4, 0x4, 0xc000406ba0, 0x4, 0x4, 0xc000a23a90, 0x4540fc, ...)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xb933a0, 0xcbd088, 0x13, 0xc000406ba0, 0x4, 0x4, 0xb970c0, 0xc00009ce10, 0xc000a23b28)
	/opt/hostedtoolcache/go/1.14.15/x64/src/reflect/value.go:321 +0xb4
github.com/openark/orchestrator/vendor/github.com/codegangsta/inject.(*injector).Invoke(0xc000e4e000, 0xb933a0, 0xcbd088, 0xc000e4e000, 0xc0000920a0, 0xc000a7e3c0, 0xc000e4e000, 0xc00009ce10)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/codegangsta/inject/inject.go:102 +0x2e3
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*context).run(0xc0000920a0)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:173 +0x71
github.com/openark/orchestrator/vendor/github.com/go-martini/martini.(*Martini).ServeHTTP(0xc000115400, 0xda7280, 0xc000118000, 0xc000139700)
	/home/runner/work/orchestrator/orchestrator/.gopath/src/github.com/openark/orchestrator/vendor/github.com/go-martini/martini/martini.go:69 +0x57
net/http.serverHandler.ServeHTTP(0xc0001181c0, 0xda7280, 0xc000118000, 0xc000139700)
	/opt/hostedtoolcache/go/1.14.15/x64/src/net/http/server.go:2836 +0xa3
net/http.(*conn).serve(0xc000238b40, 0xdaa9c0, 0xc000292b40)
	/opt/hostedtoolcache/go/1.14.15/x64/src/net/http/server.go:1924 +0x86c
created by net/http.(*Server).Serve
	/opt/hostedtoolcache/go/1.14.15/x64/src/net/http/server.go:2962 +0x35c
2021-06-30 16:23:23 INFO topology_recovery: Running 1 PreGracefulTakeoverProcesses hooks
2021-06-30 16:23:23 INFO topology_recovery: Running PreGracefulTakeoverProcesses hook 1 of 1: echo 'Planned takeover about to take place on mariadb-13306:3306. Master will switch to read_only' >> /tmp/recovery.log
2021-06-30 16:23:23 INFO CommandRun(echo 'Planned takeover about to take place on mariadb-13306:3306. Master will switch to read_only' >> /tmp/recovery.log,[])
2021-06-30 16:23:23 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-314300773
2021-06-30 16:23:23 INFO CommandRun: 

2021-06-30 16:23:23 INFO CommandRun successful. exit status 0
2021-06-30 16:23:23 INFO topology_recovery: Completed PreGracefulTakeoverProcesses hook 1 of 1 in 1.448539ms
2021-06-30 16:23:23 INFO topology_recovery: done running PreGracefulTakeoverProcesses hooks
2021-06-30 16:23:23 INFO GracefulMasterTakeover: Will set mariadb-13306:3306 as read_only
2021-06-30 16:23:23 INFO instance mariadb-13306:3306 read_only: true
2021-06-30 16:23:23 INFO auditType:read-only instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:set as true
2021-06-30 16:23:23 INFO GracefulMasterTakeover: Will wait for mariadb-13307:3306 to reach master coordinates mysql-bin.000003:342
2021-06-30 16:23:23 INFO GracefulMasterTakeover: attempting recovery
2021-06-30 16:23:23 INFO executeCheckAndRecoverFunction: proceeding with DeadMaster detection on mariadb-13306:3306; isActionable?: true; skipProcesses: false
2021-06-30 16:23:23 INFO topology_recovery: detected DeadMaster failure on mariadb-13306:3306
2021-06-30 16:23:23 INFO topology_recovery: Running 1 OnFailureDetectionProcesses hooks
2021-06-30 16:23:23 INFO topology_recovery: Running OnFailureDetectionProcesses hook 1 of 1: echo 'Detected DeadMaster on mariadb-13306:3306. Affected replicas: 1' >> /tmp/recovery.log
2021-06-30 16:23:23 INFO CommandRun(echo 'Detected DeadMaster on mariadb-13306:3306. Affected replicas: 1' >> /tmp/recovery.log,[])
2021-06-30 16:23:23 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-340225408
2021-06-30 16:23:23 INFO CommandRun: 

2021-06-30 16:23:23 INFO CommandRun successful. exit status 0
2021-06-30 16:23:23 INFO topology_recovery: Completed OnFailureDetectionProcesses hook 1 of 1 in 1.23278ms
2021-06-30 16:23:23 INFO topology_recovery: done running OnFailureDetectionProcesses hooks
2021-06-30 16:23:23 INFO executeCheckAndRecoverFunction: proceeding with DeadMaster recovery on mariadb-13306:3306; isRecoverable?: true; skipProcesses: false
2021-06-30 16:23:23 INFO topology_recovery: will handle DeadMaster event on mariadb-13306:3306
2021-06-30 16:23:23 INFO auditType:recover-dead-master instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:problem found; will recover
2021-06-30 16:23:23 INFO topology_recovery: Running 1 PreFailoverProcesses hooks
2021-06-30 16:23:23 INFO topology_recovery: Running PreFailoverProcesses hook 1 of 1: echo 'Will recover from DeadMaster on mariadb-13306:3306' >> /tmp/recovery.log
2021-06-30 16:23:23 INFO CommandRun(echo 'Will recover from DeadMaster on mariadb-13306:3306' >> /tmp/recovery.log,[])
2021-06-30 16:23:23 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-172375263
2021-06-30 16:23:23 INFO CommandRun: 

2021-06-30 16:23:23 INFO CommandRun successful. exit status 0
2021-06-30 16:23:23 INFO topology_recovery: Completed PreFailoverProcesses hook 1 of 1 in 1.337549ms
2021-06-30 16:23:23 INFO topology_recovery: done running PreFailoverProcesses hooks
2021-06-30 16:23:23 INFO topology_recovery: RecoverDeadMaster: will recover mariadb-13306:3306
2021-06-30 16:23:23 INFO topology_recovery: RecoverDeadMaster: masterRecoveryType=MasterRecoveryGTID
2021-06-30 16:23:23 INFO topology_recovery: RecoverDeadMaster: regrouping replicas via GTID
2021-06-30 16:23:23 DEBUG GetCandidateReplica: candidate: mariadb-13307:3306, ahead: 0, equal: 0, late: 0, break: 0
2021-06-30 16:23:23 INFO topology_recovery: RecoverDeadMaster: promotedReplicaIsIdeal(mariadb-13307:3306)
2021-06-30 16:23:23 DEBUG RegroupReplicasGTID: working on 0 replicas
2021-06-30 16:23:23 INFO Started replication on mariadb-13307:3306
2021-06-30 16:23:24 DEBUG RegroupReplicasGTID: done
2021-06-30 16:23:24 INFO auditType:regroup-replicas-gtid instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:regrouped replicas of mariadb-13306:3306 via GTID; promoted mariadb-13307:3306
2021-06-30 16:23:24 INFO auditType:begin-downtime instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:owner: lei-thinkcentre-e12, reason: lost-in-recovery
2021-06-30 16:23:24 INFO topology_recovery: RecoverDeadMaster: 1 postponed functions
2021-06-30 16:23:24 INFO topology_recovery: checking if should replace promoted replica with a better candidate
2021-06-30 16:23:24 INFO topology_recovery: + sanity check: found our very own server to promote; doing nothing
2021-06-30 16:23:24 INFO topology_recovery: replace-promoted-replica-with-candidate: promoted instance mariadb-13307:3306 requires no further action
2021-06-30 16:23:24 INFO topology_recovery: promoted replica: mariadb-13307:3306
2021-06-30 16:23:24 INFO auditType:recover-dead-master instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:promoted replica: mariadb-13307:3306
2021-06-30 16:23:24 INFO topology_recovery: RecoverDeadMaster: successfully promoted mariadb-13307:3306
2021-06-30 16:23:24 INFO topology_recovery: - RecoverDeadMaster: promoted server coordinates: mysql-bin.000003:342
2021-06-30 16:23:24 INFO topology_recovery: - RecoverDeadMaster: will apply MySQL changes to promoted master
2021-06-30 16:23:24 INFO Will reset replica on mariadb-13307:3306
2021-06-30 16:23:24 INFO auditType:begin-maintenance instance:mariadb-13307:3306 cluster:mariadb-13306:3306 message:maintenanceToken: 2, owner: lei-thinkcentre-e12, reason: reset replica
2021-06-30 16:23:24 INFO Stopped replication on mariadb-13307:3306, Self:mysql-bin.000003:342, Exec:mysql-bin.000003:342
2021-06-30 16:23:24 INFO Reset replication mariadb-13307:3306
2021-06-30 16:23:24 INFO auditType:reset-slave instance:mariadb-13307:3306 cluster:mariadb-13306:3306 message:mariadb-13307:3306 replication reset
2021-06-30 16:23:24 INFO auditType:end-maintenance instance:mariadb-13307:3306 cluster:mariadb-13306:3306 message:maintenanceToken: 2
2021-06-30 16:23:24 INFO topology_recovery: - RecoverDeadMaster: applying RESET SLAVE ALL on promoted master: success=true
2021-06-30 16:23:24 INFO instance mariadb-13307:3306 read_only: false
2021-06-30 16:23:24 INFO auditType:read-only instance:mariadb-13307:3306 cluster:mariadb-13306:3306 message:set as false
2021-06-30 16:23:24 INFO topology_recovery: - RecoverDeadMaster: applying read-only=0 on promoted master: success=true
2021-06-30 16:23:24 INFO topology_recovery: Writing KV [mysql/master/mariadb-13306:mariadb-13307:3306 mysql/master/mariadb-13306/hostname:mariadb-13307 mysql/master/mariadb-13306/port:3306 mysql/master/mariadb-13306/ipv4:172.31.254.2 mysql/master/mariadb-13306/ipv6:]
2021-06-30 16:23:24 INFO topology_recovery: Distributing KV [mysql/master/mariadb-13306:mariadb-13307:3306 mysql/master/mariadb-13306/hostname:mariadb-13307 mysql/master/mariadb-13306/port:3306 mysql/master/mariadb-13306/ipv4:172.31.254.2 mysql/master/mariadb-13306/ipv6:]
2021-06-30 16:23:24 INFO topology_recovery: - RecoverDeadMaster: updating cluster_alias: mariadb-13306:3306 -> mariadb-13307:3306
2021-06-30 16:23:24 INFO topology_recovery: Running 1 PostMasterFailoverProcesses hooks
2021-06-30 16:23:24 INFO topology_recovery: Running PostMasterFailoverProcesses hook 1 of 1: echo 'Recovered from DeadMaster on mariadb-13306:3306. Failed: mariadb-13306:3306; Promoted: mariadb-13307:3306' >> /tmp/recovery.log
2021-06-30 16:23:24 INFO CommandRun(echo 'Recovered from DeadMaster on mariadb-13306:3306. Failed: mariadb-13306:3306; Promoted: mariadb-13307:3306' >> /tmp/recovery.log,[])
2021-06-30 16:23:24 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-492840626
2021-06-30 16:23:24 INFO CommandRun: 

2021-06-30 16:23:24 INFO CommandRun successful. exit status 0
2021-06-30 16:23:24 INFO topology_recovery: Completed PostMasterFailoverProcesses hook 1 of 1 in 2.339314ms
2021-06-30 16:23:24 INFO topology_recovery: done running PostMasterFailoverProcesses hooks
2021-06-30 16:23:24 INFO Topology recovery: {"Id":1,"UID":"1625041403983287271:14271f9af1e540269ad1261d356445037e66f4bb09c1b224495f40a90949821e","AnalysisEntry":{"AnalyzedInstanceKey":{"Hostname":"mariadb-13306","Port":3306},"AnalyzedInstanceMasterKey":{"Hostname":"","Port":0},"ClusterDetails":{"ClusterName":"mariadb-13306:3306","ClusterAlias":"mariadb-13306","ClusterDomain":"","CountInstances":3,"HeuristicLag":0,"HasAutomatedMasterRecovery":false,"HasAutomatedIntermediateMasterRecovery":false},"AnalyzedInstanceDataCenter":"","AnalyzedInstanceRegion":"","AnalyzedInstancePhysicalEnvironment":"","AnalyzedInstanceBinlogCoordinates":{"LogFile":"mysql-bin.000003","LogPos":342,"Type":0},"IsMaster":true,"IsReplicationGroupMember":false,"IsCoMaster":false,"LastCheckValid":true,"LastCheckPartialSuccess":true,"CountReplicas":1,"CountValidReplicas":1,"CountValidReplicatingReplicas":1,"CountReplicasFailingToConnectToMaster":0,"CountDowntimedReplicas":0,"ReplicationDepth":0,"Replicas":[{"Hostname":"mariadb-13307","Port":3306}],"SlaveHosts":[{"Hostname":"mariadb-13307","Port":3306}],"IsFailingToConnectToMaster":false,"Analysis":"DeadMaster","Description":"","StructureAnalysis":null,"IsDowntimed":false,"IsReplicasDowntimed":false,"DowntimeEndTimestamp":"","DowntimeRemainingSeconds":0,"IsBinlogServer":false,"PseudoGTIDImmediateTopology":false,"OracleGTIDImmediateTopology":false,"MariaDBGTIDImmediateTopology":true,"BinlogServerImmediateTopology":false,"SemiSyncMasterEnabled":true,"SemiSyncMasterStatus":true,"SemiSyncMasterWaitForReplicaCount":0,"SemiSyncMasterClients":2,"CountSemiSyncReplicasEnabled":1,"CountLoggingReplicas":1,"CountStatementBasedLoggingReplicas":0,"CountMixedBasedLoggingReplicas":0,"CountRowBasedLoggingReplicas":1,"CountDistinctMajorVersionsLoggingReplicas":1,"CountDelayedReplicas":0,"CountLaggingReplicas":0,"IsActionableRecovery":true,"ProcessingNodeHostname":"lei-thinkcentre-e12","ProcessingNodeToken":"f63123c164eccae398c39b27201aafb7ac34996afd3b6982691204d3d05848cf","CountAdditionalAgreeingNodes":0,"StartActivePeriod":"","SkippableDueToDowntime":false,"GTIDMode":"","MinReplicaGTIDMode":"","MaxReplicaGTIDMode":"","MaxReplicaGTIDErrant":"","CommandHint":"graceful-master-takeover","IsReadOnly":false},"SuccessorKey":{"Hostname":"mariadb-13307","Port":3306},"SuccessorAlias":"","SuccessorBinlogCoordinates":{"LogFile":"mysql-bin.000003","LogPos":342,"Type":0},"IsActive":false,"IsSuccessful":true,"LostReplicas":[],"ParticipatingInstanceKeys":[],"AllErrors":[],"RecoveryStartTimestamp":"","RecoveryEndTimestamp":"","ProcessingNodeHostname":"","ProcessingNodeToken":"","Acknowledged":false,"AcknowledgedAt":"","AcknowledgedBy":"","AcknowledgedComment":"","LastDetectionId":0,"RelatedRecoveryId":0,"Type":"MasterRecovery","RecoveryType":"MasterRecoveryGTID"}
2021-06-30 16:23:24 INFO topology_recovery: Running 1 PostFailoverProcesses hooks
2021-06-30 16:23:24 INFO topology_recovery: Running PostFailoverProcesses hook 1 of 1: echo '(for all types) Recovered from DeadMaster on mariadb-13306:3306. Failed: mariadb-13306:3306; Successor: mariadb-13307:3306' >> /tmp/recovery.log
2021-06-30 16:23:24 INFO CommandRun(echo '(for all types) Recovered from DeadMaster on mariadb-13306:3306. Failed: mariadb-13306:3306; Successor: mariadb-13307:3306' >> /tmp/recovery.log,[])
2021-06-30 16:23:24 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-508393577
2021-06-30 16:23:24 INFO CommandRun: 

2021-06-30 16:23:24 INFO CommandRun successful. exit status 0
2021-06-30 16:23:24 INFO topology_recovery: Completed PostFailoverProcesses hook 1 of 1 in 2.37118ms
2021-06-30 16:23:24 INFO topology_recovery: done running PostFailoverProcesses hooks
2021-06-30 16:23:24 INFO topology_recovery: Waiting for 1 postponed functions
2021-06-30 16:23:24 DEBUG PostponedFunctionsContainer: waiting on 1 postponed functions
2021-06-30 16:23:24 DEBUG PostponedFunctionsContainer: done waiting
2021-06-30 16:23:24 INFO topology_recovery: Executed 1 postponed functions
2021-06-30 16:23:24 INFO topology_recovery: Executed postponed functions: regroup-replicas-gtid mariadb-13307:3306
2021-06-30 16:23:24 INFO instance mariadb-13306:3306 read_only: true
2021-06-30 16:23:24 INFO auditType:read-only instance:mariadb-13306:3306 cluster:mariadb-13306:3306 message:set as true
2021-06-30 16:23:24 INFO topology_recovery: - RecoverDeadMaster: applying read-only=1 on demoted master: success=true
2021-06-30 16:23:24 DEBUG ChangeMasterTo: will attempt changing master on mariadb-13306:3306 to mariadb-13307:3306, mysql-bin.000003:342
2021-06-30 16:23:24 INFO ChangeMasterTo: Changed master on mariadb-13306:3306 to: mariadb-13307:3306, mysql-bin.000003:342. GTID: true
2021-06-30 16:23:24 INFO topology_recovery: Running 1 PostGracefulTakeoverProcesses hooks
2021-06-30 16:23:24 INFO topology_recovery: Running PostGracefulTakeoverProcesses hook 1 of 1: echo 'Planned takeover complete' >> /tmp/recovery.log
2021-06-30 16:23:24 INFO CommandRun(echo 'Planned takeover complete' >> /tmp/recovery.log,[])
2021-06-30 16:23:24 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-929831092
2021-06-30 16:23:24 INFO CommandRun: 

2021-06-30 16:23:24 INFO CommandRun successful. exit status 0
2021-06-30 16:23:24 INFO topology_recovery: Completed PostGracefulTakeoverProcesses hook 1 of 1 in 1.368904ms
2021-06-30 16:23:24 INFO topology_recovery: done running PostGracefulTakeoverProcesses hooks
[martini] Completed 200 OK in 272.313085ms
[martini] Started GET /web/cluster/alias/mariadb-13306 for 127.0.0.1:59466
[martini] Completed 200 OK in 3.568343ms
[martini] Started GET /api/clusters-info for 127.0.0.1:59466
[martini] Started GET /api/check-global-recoveries for 127.0.0.1:59474
[martini] Completed 200 OK in 1.795768ms
[martini] Completed 200 OK in 1.597506ms
[martini] Started GET /api/cluster/mariadb-13307:3306 for 127.0.0.1:59474
[martini] Started GET /api/active-cluster-recovery/mariadb-13307:3306 for 127.0.0.1:59470
[martini] Started GET /api/cluster-info/mariadb-13307:3306 for 127.0.0.1:59466
[martini] Started GET /api/recently-active-cluster-recovery/mariadb-13307:3306 for 127.0.0.1:59418
[martini] Completed 200 OK in 6.711212ms
[martini] Completed 200 OK in 5.060261ms
[martini] Completed 200 OK in 3.595343ms
[martini] Completed 200 OK in 7.091433ms
[martini] Started GET /api/blocked-recoveries/cluster/mariadb-13307:3306 for 127.0.0.1:59420
[martini] Started GET /api/problems/mariadb-13307:3306 for 127.0.0.1:59466
[martini] Completed 200 OK in 2.253599ms
[martini] Completed 200 OK in 2.873622ms
[martini] Started GET /api/replication-analysis/mariadb-13307:3306 for 127.0.0.1:59466
[martini] Completed 200 OK in 10.013202ms
[martini] Started GET /api/audit-recovery/alias/mariadb-13306 for 127.0.0.1:59466
[martini] Started GET /api/audit-failure-detection/alias/mariadb-13306 for 127.0.0.1:59420
[martini] Completed 200 OK in 2.08584ms
[martini] Completed 200 OK in 2.370572ms
[martini] Started GET /api/maintenance for 127.0.0.1:59420
[martini] Completed 200 OK in 1.017607ms
[martini] Started GET /api/maintenance for 127.0.0.1:59420
[martini] Completed 200 OK in 886.226µs
[martini] Started GET /api/recently-active-instance-recovery/mariadb-13307/3306 for 127.0.0.1:59420
[martini] Completed 200 OK in 2.32736ms
[martini] Started GET / for 127.0.0.1:59420
[martini] Completed 302 Found in 671.469µs
[martini] Started GET /web/clusters for 127.0.0.1:59420
[martini] Completed 200 OK in 2.220419ms
[martini] Started GET /api/clusters-info for 127.0.0.1:59420
[martini] Started GET /api/check-global-recoveries for 127.0.0.1:59466
[martini] Completed 200 OK in 1.52547ms
[martini] Completed 200 OK in 1.162033ms
[martini] Started GET /api/clusters-info for 127.0.0.1:59466
[martini] Started GET /api/problems for 127.0.0.1:59420
[martini] Completed 200 OK in 1.89455ms
[martini] Completed 200 OK in 3.556486ms
[martini] Started GET /api/replication-analysis for 127.0.0.1:59420
[martini] Started GET /api/maintenance for 127.0.0.1:59466
[martini] Completed 200 OK in 1.142963ms
[martini] Completed 200 OK in 10.204643ms
[martini] Started GET /api/problems for 127.0.0.1:59420
[martini] Completed 200 OK in 2.817782ms
[martini] Started GET /web/cluster/alias/mariadb-13307 for 127.0.0.1:59420
[martini] Completed 200 OK in 2.385468ms
[martini] Started GET /api/clusters-info for 127.0.0.1:59420
[martini] Started GET /api/check-global-recoveries for 127.0.0.1:59466
[martini] Completed 200 OK in 2.685524ms
[martini] Completed 200 OK in 3.647023ms
[martini] Started GET /api/cluster/mariadb-13307:3306 for 127.0.0.1:59466
[martini] Started GET /api/cluster-info/mariadb-13307:3306 for 127.0.0.1:59420
[martini] Started GET /api/active-cluster-recovery/mariadb-13307:3306 for 127.0.0.1:59418
[martini] Started GET /api/recently-active-cluster-recovery/mariadb-13307:3306 for 127.0.0.1:59470
[martini] Started GET /api/blocked-recoveries/cluster/mariadb-13307:3306 for 127.0.0.1:59474
[martini] Completed 200 OK in 1.670547ms
[martini] Completed 200 OK in 1.192136ms
[martini] Completed 200 OK in 1.350641ms
[martini] Completed 200 OK in 1.572593ms
[martini] Completed 200 OK in 3.514869ms
[martini] Started GET /api/problems/mariadb-13307:3306 for 127.0.0.1:59470
[martini] Completed 200 OK in 4.030063ms
[martini] Started GET /api/audit-recovery/alias/mariadb-13307 for 127.0.0.1:59470
[martini] Started GET /api/audit-failure-detection/alias/mariadb-13307 for 127.0.0.1:59466
[martini] Completed 200 OK in 1.959578ms
[martini] Completed 200 OK in 2.058022ms
[martini] Started GET /api/replication-analysis/mariadb-13307:3306 for 127.0.0.1:59470
[martini] Started GET /api/maintenance for 127.0.0.1:59466
[martini] Completed 200 OK in 799.082µs
[martini] Completed 200 OK in 9.402254ms
[martini] Started GET /api/maintenance for 127.0.0.1:59470
[martini] Completed 200 OK in 1.125868ms
[martini] Started GET /api/recently-active-instance-recovery/mariadb-13307/3306 for 127.0.0.1:59470
[martini] Completed 200 OK in 4.027481ms
[martini] Started GET /api/master-equivalent/mariadb-13307/3306//0 for 127.0.0.1:59470
[martini] Completed 404 Not Found in 1.294529ms

```
</details>

**Comment by d-rupp** [comment]
FYI: I built docker containers from the branch to test this and it worked fine. I did not use the build artifact. 
Also i am not using the slave_master_info table, i use a dedicated metadata table where the replication credentials are stored.
I fetch the credentials with this:
```"ReplicationCredentialsQuery": "select replication_user, replication_password from percona.cluster where anchor=1",```

**Comment by leiless** [comment]
I noticed that operation logs had an error line
```
2021-06-30 16:23:23 ERROR Error 1142: SELECT command denied to user 'orchestrator'@'172.31.254.254' for table 'slave_master_info'
```

And according to https://github.com/openark/orchestrator/blob/master/docs/install.md#grant-access-to-orchestrator-on-all-your-mysql-servers
```sql
CREATE USER 'orchestrator'@'orch_host' IDENTIFIED BY 'orch_topology_password';
GRANT SUPER, PROCESS, REPLICATION SLAVE, RELOAD ON *.* TO 'orchestrator'@'orch_host';
GRANT SELECT ON mysql.slave_master_info TO 'orchestrator'@'orch_host';
GRANT SELECT ON ndbinfo.processes TO 'orchestrator'@'orch_host'; -- Only for NDB Cluster
```

Yet MariaDB seems have not `mysql.slave_master_info` table:
```
$ mysql -h172.31.254.1 -uroot -p123456 mysql -e 'select * from slave_master_info' 
ERROR 1146 (42S02) at line 1: Table 'mysql.slave_master_info' doesn't exist

```

I don't know if it's some features needed to be configurated to be used.

According to https://mariadb.com/kb/en/mariadb-1004-changelog/
> * Revision #3492.3.49 Fri 2013-07-05 10:21:15 +0400
>   * Fix a number of tests:
>     * MariaDB does not have `mysql.slave_master_info`, `mysql.slave_relay_log_info`, `mysql.slave_worker_info` or `mysql.ndb_binlog_index` tables.
>     * Some tests expected to have these tables (this was an incorrect merge from 5.6, which merged necessary tables like `mysql.innodb*stats`, but also got these tables)

So `orchestrator` cannot grab replication credentials from `mysql.slave_master_info`.

**Comment by leiless** [comment]
> FYI: I built docker containers from the branch to test this and it worked fine. I did not use the build artifact.
> Also i am not using the slave_master_info table, i use a dedicated metadata table where the replication credentials are stored.
> I fetch the credentials with this:
> `"ReplicationCredentialsQuery": "select replication_user, replication_password from percona.cluster where anchor=1",`

Thx for your reply, I've check my `/etc/orchestrator.conf.json` file, and I found that the `ReplicationCredentialsQuery` is missing from my config(which previously I copied from `orchestrator-sample-sqlite.conf.json`).

So does that meaning I do not properly guide the `orchestrator` from where to find the replication credential so I failed `graceful-master-takeover`?

**Comment by d-rupp** [comment]
Two things: 
1) I don´t think MariaDB has the function to store the master-info-repository in a table. We are using Percona Server 5.7
2) I think that if you don´t have the table you have to tell Orchestrator where to find the credential information.

**Comment by d-rupp** [comment]
Ah, i forgot - we also do not use the function to store it in a table although we could.

**Comment by shlomi-noach** [comment]
the fact mariadb does not have info tables is fine. `orchestrator` does not require the table to exist. I fit exists, it tries to use it. `orchestrator` is aware that the table only exists in MySQL and not in MariaDB.

Yes, please configure `ReplicationCredentialsQuery`.

**Comment by leiless** [comment]
> FYI: I built docker containers from the branch to test this and it worked fine. I did not use the build artifact.
> Also i am not using the slave_master_info table, i use a dedicated metadata table where the replication credentials are stored.
> I fetch the credentials with this:
> `"ReplicationCredentialsQuery": "select replication_user, replication_password from percona.cluster where anchor=1",`

How do you insert cluster info into `percoa.cluster`? by manually insert or what?

I wonder do we need to create a separate user for replication or reuse `MySQLTopologyUser` and `MySQLTopologyPassword`, since the topology user must have `REPLICATION CLIENT` privilege.

Is there a best practice guideline? cc @d-rupp, @shlomi-noach

see also:
https://github.com/openark/orchestrator/blob/master/docs/configuration-discovery-basic.md
https://github.com/openark/orchestrator/blob/master/docs/configuration-discovery-classifying.md

**Comment by leiless** [comment]
BTW, @d-rupp, does the table schema of your `percona.cluster` is similiar as described in https://github.com/openark/orchestrator/blob/master/docs/configuration-discovery-classifying.md#cluster-alias?

e.g.
```sql
CREATE TABLE IF NOT EXISTS cluster (
  anchor TINYINT NOT NULL,
  -- ...
  replication_user VARCHAR(128) CHARSET ascii NOT NULL,
  replication_password VARCHAR(128) CHARSET ascii NOT NULL,
  -- ...
  PRIMARY KEY (anchor)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

**Comment by leiless** [comment]
FYI, I've fixed the problem after I configurated `ReplicationCredentialsQuery` in `/etc/orchestrator.conf.json`:

Utilized the `orchestrator` binary in https://github.com/openark/orchestrator/suites/3110208007/artifacts/71167636

```json
{
  "Debug": true,
  "EnableSyslog": false,
  "ListenAddress": ":3000",
  "ReplicationCredentialsQuery": "SELECT 'orchestrator', '123456'",
  "MySQLTopologyUser": "orchestrator",
  "MySQLTopologyPassword": "123456",

  "...": "..."
}
```

I reused topology user as replication credential(in testing environment), but I don't know if it's a good idea.

**Comment by shlomi-noach** [comment]
> How do you insert cluster info into percoa.cluster? by manually insert or what?

I have no idea what `percona.cluster` means. This doesn't seem to be related to `orchestrator`.

**Comment by d-rupp** [comment]
We have a database "percona" in all our systems that holds metadata, for example the heartbeat table. 

I created the cluster table and it is filled on server creation. It also holds further metadata that orchestrator uses.

```
CREATE TABLE `cluster` (
  `anchor` tinyint(4) NOT NULL,
  `cluster_name` varchar(128) CHARACTER SET ascii NOT NULL DEFAULT '',
  `cluster_domain` varchar(128) CHARACTER SET ascii NOT NULL DEFAULT '',
  `cluster_dc` varchar(128) CHARACTER SET ascii NOT NULL DEFAULT '',
  `replication_user` varchar(128) CHARACTER SET ascii NOT NULL DEFAULT '',
  `replication_password` varchar(128) CHARACTER SET ascii NOT NULL DEFAULT '',
  PRIMARY KEY (`anchor`)
)
```

Hope that helps a bit.

**Comment by shlomi-noach** [comment]
I'm confused. @d-rupp and @leiless are you colleagues? I'm further confused where this issue is going. It is now closed. Is there still a problem with the original issue, or is this an altogether different discussion? If so, let's take it elsewhere, into another issue.

**Comment by leiless** [comment]
> I'm confused. @d-rupp and @leiless are you colleagues? I'm further confused where this issue is going. It is now closed. Is there still a problem with the original issue, or is this an altogether different discussion? If so, let's take it elsewhere, into another issue.

No, we definitely not.
The original issue has been solved by now.

**Comment by d-rupp** [comment]
No, i just wanted to help confirm that it is fixed. Sorry for the noise.

**Comment by shlomi-noach** [comment]
Haha, you both have a `percona.cluster` table and were both discussing it so that confused me. Thank you!
